### PR TITLE
add stat collection objects

### PR DESCRIPTION
**Pull request by dkadiyala3**
## <u>Changes</u>:
1. Added the network io statistics collection module under extern/statistics
2. Fixed the runtime calculation in AnalyticalRemoteMemory::get_remote_mem_runtime(uint64_t tensor_size) to compute in ns.
3. Linked execution-graph-converter-Intel submodule
4. Added `inputs/workload/execution-graph-converter-Intel/conv-spreadsheet-to-exegraph.sh` script to generate the inputs from spreadsheet models into the execution graphs under `inputs/workload/ASTRA-sim-2.0/` folder
5. Added `inputs/workload/conv-text-to-exegraph.sh` script to generate the inputs from ASTRA-sim-1.0 text files into the execution graphs under `inputs/workload/ASTRA-sim-2.0/` folder

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -13,3 +13,6 @@
 [submodule "extern/network_backend/ns3"]
 	path = extern/network_backend/ns3
 	url = https://github.com/astra-sim/astra-network-ns3.git
+[submodule "inputs/workload/execution-graph-converter-Intel"]
+	path = inputs/workload/execution-graph-converter-Intel
+	url = git@github.com:astra-sim/execution-graph-converter-Intel.git
```
Comment (TaekyungHeo) [inline .gitmodules:18]: As the name suggests, we are not allowed to release this repository. We don't have any plan to make it public. Which tool do you need for this PR?

File: astra-sim/network_frontend/analytical/congestion_unaware/AnalyticalNetwork.cc [hunk 2]
```diff
@@ -4,7 +4,7 @@ the root directory of this source tree.
  *******************************************************************************/
 
 #include "AnalyticalNetwork.hh"
-
+#include "cassert"
```
Comment (TaekyungHeo) [inline astra-sim/network_frontend/analytical/congestion_unaware/AnalyticalNetwork.cc:7]: I believe you can replace it with #include <cassert>

File: astra-sim/network_frontend/analytical/congestion_unaware/AnalyticalNetwork.cc [hunk 3]
```diff
@@ -41,7 +43,7 @@ int AnalyticalNetwork::sim_send(
   auto used_dim = -1;
   std::tie(delta.time_val, used_dim) =
       topology->send(src, dst, count); // simulate src->dst and get latency
-
+                                       // 'count' here is the payload size in bytes.
```
Comment (TaekyungHeo) [inline astra-sim/network_frontend/analytical/congestion_unaware/AnalyticalNetwork.cc:46]: Thanks for adding a comment.

File: astra-sim/network_frontend/analytical/congestion_unaware/AnalyticalNetwork.cc [hunk 4]
```diff
@@ -57,6 +59,7 @@ int AnalyticalNetwork::sim_send(
         delta,
         recv_event_handler.get_fun_ptr(),
         recv_event_handler.get_fun_arg());
+
```
Comment (TaekyungHeo) [inline astra-sim/network_frontend/analytical/congestion_unaware/AnalyticalNetwork.cc:62]: Please remove this blank line.

File: astra-sim/network_frontend/analytical/congestion_unaware/AnalyticalNetwork.cc [hunk 5]
```diff
@@ -73,6 +76,29 @@ int AnalyticalNetwork::sim_send(
     send_recv_tracking_map.insert_send(tag, src, dst, count, send_finish_time);
   }
 
+  // Record the i/o ioActivity for both sender and receiver.
+  // Remember: the moment sender starts sending, receiver's ingress also
+  // gets busy. Scheduling the recev_handler is not the same as actually
+  // receiving the packet. recev_handler is to tell sys what to do once
+  // after the packet is arrived at the dest.
```
Comment (TaekyungHeo) [inline astra-sim/network_frontend/analytical/congestion_unaware/AnalyticalNetwork.cc:83]: Comments look good.

File: astra-sim/network_frontend/analytical/congestion_unaware/AnalyticalNetwork.cc [hunk 6]
```diff
@@ -111,6 +137,7 @@ int AnalyticalNetwork::sim_recv(
 
     // schedule recv handler
     sim_schedule(delta, msg_handler, fun_arg);
+
```
Comment (TaekyungHeo) [inline astra-sim/network_frontend/analytical/congestion_unaware/AnalyticalNetwork.cc:140]: Don't need to add a blank line

File: build/astra_analytical/CMakeLists.txt [hunk 7]
```diff
@@ -5,7 +5,10 @@ cmake_minimum_required(VERSION 3.15)
 set(CMAKE_CXX_STANDARD 17)
 set(CMAKE_CXX_STANDARD_REQUIRED ON)
 
-set(CMAKE_CXX_FLAGS "-O3 -Wall")
+#set(CMAKE_CXX_FLAGS "-O3 -Wall -g")
+#set(CMAKE_CXX_FLAGS "-Wall -ggdb3 -Wextra")
+#set(CMAKE_CXX_FLAGS "-Wall -ggdb3 -O3 -Wextra -pedantic -fsanitize=address")
```
Comment (TaekyungHeo) [inline build/astra_analytical/CMakeLists.txt:10]: Please avoid adding commented out lines. Instead, you may introduce switches to Cmake.

File: build/astra_analytical/build.sh [hunk 8]
```diff
@@ -25,8 +25,10 @@ function setup {
 function compile {
   # compile AstraSim
   cd "${BUILD_DIR:?}" || exit
-  cmake ..
-  cmake --build . -j "${NUM_THREADS:?}"
+  #cmake ..
+  #cmake --build . -j "${NUM_THREADS:?}"
```
Comment (TaekyungHeo) [inline build/astra_analytical/build.sh:29]: Please remove commented out lines

File: extern/statistics/GStats.cc [hunk 9]
```diff
@@ -0,0 +1,95 @@
+//
+// Created by dkadiyala3 on 10/29/23.
+//
+
+#include "GStats.hh"
```
Comment (TaekyungHeo) [inline extern/statistics/GStats.cc:5]: Please add a blank line below this line.

File: .gitignore [hunk 10]
```diff
@@ -14,3 +14,9 @@ build/astra_analytical/build/
 build/astra_analytical/result/
 build/astra_garnet/build/
 build/astra_garnet/result/
+
+# Execution graphs
+inputs/workload/ASTRA-sim-2.0/*.eg
```
Comment (TaekyungHeo) [inline .gitignore:19]: Please add `inputs/workload/ASTRA-sim-2.0/*.et` as well

File: extern/statistics/GStats.cc [hunk 11]
```diff
@@ -0,0 +1,95 @@
+//
+// Created by dkadiyala3 on 10/29/23.
+//
```
Comment (TaekyungHeo) [inline extern/statistics/GStats.cc:3]: You may need to add a license block instead.

File: extern/statistics/GStats.cc [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+    if(*i == this) {
+      store->erase(i);
+    }
+  }
+}
+
+void GStats::report(void)
+{
+  //Report::flush();
+  //Report::field("BEGIN GStats::report %s", str);
```
Comment (TaekyungHeo) [inline extern/statistics/GStats.cc:26]: Please remove commented-out codes.

File: extern/statistics/GStats.cc [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+  // Get the iterator the current position in the list.
+  cpos=store->emplace(store->end(),this);
+  // Later add a message mentioning a container has been sunscribed.
+}
+
+void GStats::unsubscribe(void)
+{
+  assert(store);
+  assert(*cpos==this);
+  //store->erase(cpos);
```
Comment (beckmabd) [inline extern/statistics/GStats.cc:53]: please remove commented out code here and the couple lines later.

File: extern/statistics/GStats.hh [hunk 14]
```diff
@@ -0,0 +1,55 @@
+//
```
Comment (beckmabd) [inline extern/statistics/GStats.hh:1]: license and copyright block

File: extern/statistics/GStatsIoActivity.cc [hunk 15]
```diff
@@ -0,0 +1,108 @@
+//
```
Comment (beckmabd) [inline extern/statistics/GStatsIoActivity.cc:1]: license and copyright

File: extern/statistics/GStatsIoActivity.hh [hunk 16]
```diff
@@ -0,0 +1,35 @@
+//
```
Comment (beckmabd) [inline extern/statistics/GStatsIoActivity.hh:1]: license and copyright

**Review by beckmabd — CHANGES_REQUESTED** [review]

**Comment by willjwon** [comment]
@dkadiyala3 I think this PR will experience merge conflicts with the latest main branch, due to the Network API changes in the analytical backend 2.0. I'll look into this and try creating a patch branch if possible. Thanks!

**Comment by jinsun-yoo** [comment]
Close this PR for now due to updates in network backend.

**Pull request closed without merging**